features/atcryptoauth/mbedtls_atca_engine.cpp

```diff
@@ -0,0 +1,355 @@
+
+    if( ( ret = mbedtls_asn1_get_tag( &p, end, &len,
+                    MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE ) ) != 0 )
+    {
+        ret += MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
+        goto cleanup;
+    }
+
+    if( p + len != end )
+    {
+        ret = MBEDTLS_ERR_ECP_BAD_INPUT_DATA +
+              MBEDTLS_ERR_ASN1_LENGTH_MISMATCH;
+        goto cleanup;
+    }
+
+    if( ( ret = mbedtls_asn1_get_mpi( &p, end, &r ) ) != 0 ||
+        ( ret = mbedtls_asn1_get_mpi( &p, end, &s ) ) != 0 )
+    {
+        ret += MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
+        goto cleanup;
+    }
+
+    if( ( ret = mbedtls_mpi_write_binary( &r, R, R_len ) ) != 0 ||
+        ( ret = mbedtls_mpi_write_binary( &s, S, S_len ) ) != 0 )
+    {
+        ret += MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
+        goto cleanup;
+    }
+
+    if( p != end )
+        ret = MBEDTLS_ERR_ECP_SIG_LEN_MISMATCH;
+
+cleanup:
+    mbedtls_mpi_free( &r );
+    mbedtls_mpi_free( &s );
+
+    return( ret );
+}
+
+/** Verify signature.
+ *
+ *
+ * @param ctx       Key context
+ * @param md_alg    Algorithm that was used to hash the message
+ * @param hash      Message hash
+ * @param hash_len  Length of hash
+ * @param sig       Signature buffer
+ * @param sig_len   Length of the signature buffer
+ *
+ * @retval          0 if successful, or appropriate Mbed TLS error code.
+ */
+static int atca_verify_func( void *ctx, mbedtls_md_type_t md_alg,
+                        const unsigned char *hash, size_t hash_len,
+                        const unsigned char *sig, size_t sig_len )
+{
+    ATCAKey * key = (ATCAKey *)ctx;
+    uint8_t rs[64];
+    int ret = 0;
+
+    if ( md_alg != MBEDTLS_MD_SHA256 )
+        return MBEDTLS_ERR_MD_FEATURE_UNAVAILABLE;
+
+    /* Get R & S concatenantion from signature */
+    if ( (ret = mbedtls_ecdsa_asn1_to_signature(sig, sig_len, rs, sizeof(rs)/2,
+                    rs + sizeof(rs)/2, sizeof(rs)/2)) != 0)
+    {
+        printf("R & S import failed\r\n");
+        return ret;
+    }
+    /* Verify the signature */
+    ATCAError err = key->Verify( rs, sizeof(rs), hash, hash_len);
+    if (err == ATCA_ERR_CHECK_MAC_OR_VERIFY_FAIL)
+    {
+        return MBEDTLS_ERR_PK_INVALID_SIGNATURE;
+    }
+    else if (err != ATCA_SUCCESS)
+    {
+        printf("Verify failed = 0x%x\r\n", err);
+        return MBEDTLS_ERR_PK_HW_ACCEL_FAILED;
+    }
+    return 0;
+}
+
+/** Cleanup Key context.
+ *
+ * @param ctx       Key context
+ */
+static void atca_ctx_free( void * ctx )
+{
+    ATCAKey * key = (ATCAKey *)ctx;
+    delete key;
+}
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** PK Setup function for ATCA crypto engine.
+ *
+ * @param ctx       Key context
+ * @param keyId     Private key Id representing a hardware Key.
+ *
+ * @retval          0 if successful, or appropriate Mbed TLS error code.
+ */
+int mbedtls_atca_pk_setup( mbedtls_pk_context * ctx, ATCAKeyID keyId )
+{
+    ATCAKey * key = NULL;
+    ATCAError err = ATCA_SUCCESS;
```

Comment:
Same comment as [here](https://github.com/ARMmbed/mbed-os/pull/6104/files#r169607151).

Reply:
```err``` is not the return code here.